modules/kvtagger/kvtagger.c

```diff
@@ -84,10 +85,48 @@ kvtagger_set_database_selector_template(LogParser *p, const gchar *selector)
   log_template_compile(self->selector_template, selector, NULL);
 }
 
-static inline element_range const *
-kvtagger_lookup_tag(KVTagger *const self, const gchar *const key)
+void
+kvtagger_set_database_default_selector(LogParser *p, const gchar *default_selector)
 {
-  return g_hash_table_lookup(self->tagdb.index, key);
+  KVTagger *self = (KVTagger *)p;
+
+  g_free(self->default_selector);
+  self->default_selector = g_strdup(default_selector);
+}
+
+static inline element_range *
+kvtagger_lookup_tag(KVTagger *const self, const gchar *const selector)
+{
+  return g_hash_table_lookup(self->tagdb.index, selector);
+}
+
+gboolean
+_is_default_selector_set(KVTagger *self)
+{
+  return (self->default_selector != NULL);
+}
+
+element_range *
+_get_range_of_tags(KVTagger *self, GString *selector)
+{
+  element_range *position_of_tags_to_be_inserted = kvtagger_lookup_tag(self, selector->str);
+  if (position_of_tags_to_be_inserted == NULL && _is_default_selector_set(self))
+    {
+      position_of_tags_to_be_inserted = kvtagger_lookup_tag(self, self->default_selector);
+    }
+  return position_of_tags_to_be_inserted;
+}
+
+void
+_tag_message(KVTagger *self, LogMessage *msg, element_range *position_of_tags_to_be_inserted)
+{
+  for (gsize i = position_of_tags_to_be_inserted->offset;
+       i < position_of_tags_to_be_inserted->offset + position_of_tags_to_be_inserted->length;
+       ++i)
+    {
+      tag_record matching_tag = g_array_index(self->tagdb.data, tag_record, i);
+      log_msg_set_value_by_name(msg, matching_tag.name, matching_tag.value, -1);
```

Comment:
Loading should store the string length here as well, so we wouldn't need to call `strlen()` all the time due to `-1`.